java-checks/src/main/java/org/sonar/java/checks/ObjectFinalizeOverridenCallsSuperFinalizeCheck.java

```diff
@@ -47,76 +49,102 @@ public class ObjectFinalizeOverridenCallsSuperFinalizeCheck extends IssuableSubs
   private static final MethodMatchers FINALIZE_MATCHER = MethodMatchers.create()
     .ofAnyType().names(FINALIZE).addWithoutParametersMatcher().build();
 
-  private MethodInvocationTree lastStatementTree;
-
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return Arrays.asList(Tree.Kind.METHOD, Tree.Kind.METHOD_INVOCATION);
+    return List.of(Tree.Kind.METHOD);
   }
 
   @Override
   public void visitNode(Tree tree) {
-    if (tree.is(Tree.Kind.METHOD_INVOCATION)) {
-      MethodInvocationTree methodInvocationTree = (MethodInvocationTree) tree;
-      if (methodInvocationTree.methodSelect().is(Tree.Kind.MEMBER_SELECT)) {
-        MemberSelectExpressionTree mset = (MemberSelectExpressionTree) methodInvocationTree.methodSelect();
-        if (FINALIZE.equals(mset.identifier().name()) && mset.expression().is(Tree.Kind.IDENTIFIER) && "super".equals(((IdentifierTree) mset.expression()).name())) {
-          lastStatementTree = methodInvocationTree;
+
+    final MethodTree methodTree = (MethodTree) tree;
+    if (isFinalizeOverriddenMethod(methodTree)) {
+      final BlockTree blockTree = methodTree.block();
+      if (blockTree != null) {
+        final MethodInvocationTree lastSuperFinalizeInvocation = findLastSuperFinalizeInvocation(blockTree);
+        if (lastSuperFinalizeInvocation == null) {
+          reportIssue(methodTree.simpleName(), "Add a call to super.finalize() at the end of this Object.finalize() implementation.");
+        } else if (!isLastStatement(blockTree, lastSuperFinalizeInvocation)) {
+          reportIssue(lastSuperFinalizeInvocation, "Move this super.finalize() call to the end of this Object.finalize() implementation.");
         }
       }
-    } else if (FINALIZE_MATCHER.matches((MethodTree) tree)) {
-      lastStatementTree = null;
     }
   }
 
-  @Override
-  public void leaveNode(Tree tree) {
-    if (tree.is(Tree.Kind.METHOD)) {
-      MethodTree methodTree = (MethodTree) tree;
-      if (FINALIZE_MATCHER.matches(methodTree) && doesOverrideFinalize(methodTree.symbol().owner())) {
-        if (lastStatementTree == null) {
-          reportIssue(methodTree.simpleName(), "Add a call to super.finalize() at the end of this Object.finalize() implementation.");
-        } else if (!isLastStatement(methodTree, lastStatementTree)) {
-          reportIssue(lastStatementTree, "Move this super.finalize() call to the end of this Object.finalize() implementation.");
+  private static boolean isFinalizeOverriddenMethod(MethodTree methodTree) {
+    return FINALIZE_MATCHER.matches(methodTree) && doesOverrideFinalize(methodTree.symbol().owner());
+  }
+
+  private static boolean doesOverrideFinalize(@Nullable Symbol classSymbol) {
+    if ((classSymbol != null) && (classSymbol.isTypeSymbol())) {
+      Type superClassType = ((Symbol.TypeSymbol) classSymbol).superClass();
```

Comment:
As this method is only called using owners of methods, and because Java does not allow methods to be defined outside of class-like constructs, we can safely assume that the owner is never and is always a type symbol.
```suggestion
  private static boolean doesOverrideFinalize(Symbol.TypeSymbol superClassType) {
```

To validate this, you can add a new test by copying part of your test cases to a file in the `src/main/files/non-compiling/checks` and the compiler should still be able to get a really symbol and resolve the method owner as class-like thing.

Reply:
In general, I would recommend avoiding extra parentheses that don't enforce any additional precedence in if conditions by breaking the condition on multiple lines or extracting the test in a separate method